File: 0047. Permutations II.js
const permuteUnique = nums => {
  const permutations = [];

  const go = (cur, rest) => {
    if (!rest.length) {
      permutations.push(cur);
      return;
    }

    for (let i = 0; i < rest.length; i++) {
      if (rest[i] !== rest[i - 1])
        go([...cur, rest[i]], [...rest.slice(0, i), ...rest.slice(i + 1)]);
    }
  };

  go([], nums.sort());

  return permutations;
};

console.log(permuteUnique([1, 1, 2]));

File: python/0047. Permutations II.py
from typing import List
import copy


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        perms = []
        nums.sort()

        self.permute(perms, [], nums)

        return perms

    def permute(self, perms, curr: List[int], rest: List[int]) -> None:
        if len(rest) == 0:
            perms.append(copy.copy(curr))
            return None

        prev_num = float('-inf')

        for idx, num in enumerate(rest, 0):
            if num != prev_num:
                curr.append(num)

                self.permute(perms, curr, rest[:idx] + rest[idx + 1:])

                curr.pop()

            prev_num = num
